### PR TITLE
1822 character interface with llm rag web services

### PR DESCRIPTION
**Pull request by AKissmehl**

### CODE REVIEW

**Review by fwild — CHANGES_REQUESTED** [review]
The config should come from assets/resources/miragexrconfig.txt - maybe an opportunity to change that to JSON as well!

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+        /// <summary>
+        /// Provides multiple Transcription models.
+        /// </summary>
+        /// <param name="audioClip"> The audio clip that should be transcribe</param>
+        /// <param name="model"> The Model that should transcribe the audio</param>
+        /// <returns>A String with the result of task or an error if an network error appears</returns>
+        public async Task<string> Listen(AudioClip audioClip, string model)
+        {
+            UnityEngine.Debug.LogError("Speak");
```
Comment (cawke) [inline Assets/MirageXR/Scripts/AiServices/AiServices.cs]: only for debug?

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+            return webRequest.downloadHandler.text;
+        }
+
+        /// <summary>
+        /// Downloads the available option from the Server.
+        /// </summary>
+        /// <returns>A boolean. True if the operations was successful, falls if it wasn't.</returns>
+        private async Task<bool> GetOptions()
+        {
+            UnityEngine.Debug.LogError("GetOptions");
```
Comment (cawke) [inline Assets/MirageXR/Scripts/AiServices/AiServices.cs]: remove

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        /// </summary>
+        /// <param name="speakOut"> That is the text that you want to turn in to an audio</param>
+        /// <param name="voice">The voice that you wan to use. Check options json for legal parameters </param>
+        /// <param name="model">The model that use. Check options json for legal parameters</param>
+        /// <param name="onSuccess">Function that get invoke onSuccess</param>
+        /// <param name="onError">Well if it dosed work, we do this. </param>
+        /// <returns><placeholder>A <see cref="Tasks.Task"/> representing the asynchronous operation.</placeholder></returns>
+        public async Task<AudioClip> Speak(string speakOut, string voice, string model)
+        {
+            UnityEngine.Debug.LogError("Speak");
```
Comment (cawke) [inline Assets/MirageXR/Scripts/AiServices/AiServices.cs]: remove

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+                    $"API_URL = {apiURL}," +
+                    $"Username = {username}," +
+                    $"Password = {password}," +
+                    $"Token = {token}");
+            }
+            UnityEngine.Debug.LogError(apiURL+ username+ password);
+            var r = new AiServicesConfig(apiURL, username, password);
+            r.SetToken(token);
+            _config = r;
+            UnityEngine.Debug.LogError("Config is done with " + _config.AiToken + _config.AiApiUrl + _config.AiPassword + _config.AiUsername);
```
Comment (cawke) [inline Assets/MirageXR/Scripts/AiServices/AiServices.cs]: remove

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+                throw new InvalidOperationException(
+                    $"Can't find a parameter for the AI Serves configuration -> " +
+                    $"Path = {configFileName}, " +
+                    $"API_URL = {apiURL}," +
+                    $"Username = {username}," +
+                    $"Password = {password}," +
+                    $"Token = {token}");
+            }
+            UnityEngine.Debug.LogError(apiURL+ username+ password);
+            var r = new AiServicesConfig(apiURL, username, password);
```
Comment (cawke) [inline Assets/MirageXR/Scripts/AiServices/AiServices.cs]:
```suggestion
            var config = new AiServicesConfig(apiURL, username, password);
```

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+                new MultipartFormDataSection("username", user),
+                new MultipartFormDataSection("password", pass),
+            };
+            var request = UnityWebRequest.Post(apiURL, formData);
+            await request.SendWebRequest();
+            if (request.result != UnityWebRequest.Result.Success)
+            {
+                throw new HttpRequestException($"Error while receiving the token: {request.error}!");
+            }
+            var r = request.downloadHandler.text;
```
Comment (cawke) [inline Assets/MirageXR/Scripts/AiServices/AiServices.cs]:
```suggestion
            var text = request.downloadHandler.text;
```

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+            r = r.Substring(10, r.Length - 10 - 2);
+            return r;
+        }
+
+        /// <summary>
+        /// Stores the Toke for the API
+        /// </summary>
+        private readonly struct TokenResponse // still needed? @todo..
+        {
+            public TokenResponse(string t)
```
Comment (cawke) [inline Assets/MirageXR/Scripts/AiServices/AiServices.cs]:
```suggestion
            public TokenResponse(string token)
```

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+        }
+
+        /// <summary> DONE!
+        /// Get the token
+        /// </summary>
+        /// <param name="apiURL"> URL of the Server</param>
+        /// <param name="user"> Username</param>
+        /// <param name="pass"> Password</param>
+        /// <returns>The token as struct</returns>
+        private static async Task<string> AuthenticateUser(string apiURL, string user, string pass)
```
Comment (cawke) [inline Assets/MirageXR/Scripts/AiServices/AiServices.cs]:
```suggestion
        private static async Task<string> AuthenticateUser(string apiURL, string user, string password)
```

File: Assets/StreamingAssets/MirageXRConfig.txt [hunk 9]
```diff
@@ -0,0 +1,26 @@
+companyName:WEKIT ECS
```
Comment (cawke) [inline Assets/StreamingAssets/MirageXRConfig.txt]: is this an old file?

File: Assets/Resources/MirageXRConfig.txt [hunk 10]
```diff
@@ -0,0 +1,17 @@
+companyName:WEKIT ECS
```
Comment (cawke) [inline Assets/Resources/MirageXRConfig.txt]: is this an old file?

File: Assets/StreamingAssets/config.json [hunk 11]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (cawke) [inline Assets/StreamingAssets/config.json]: is this an old file?

File: Assets/Resources/MirageXRConfig.json [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+  "secondaryColor":"#FF9F1CFF",
+  "textColor":"#FFFFFFFF",
+  "iconColor":"#FFFFFFC8",
+  "taskStationColor":"#FF9F1CFF",
+  "pathColor":"#FFFFFFB5",
+  "nextPathColor":"#00FFE7C8",
+  "calibrationMarker":"Assets/Editor/Vuforia/ImageTargetTextures/MirageXRTargets/target_scaled.jpg",
+  "pdfCalibrationMarker":"Assets/MirageXR/Config/Resources/CalibrationImage/MirageXR_calibration_image_pdf.bytes",
+  "AiApiUrl": "http://127.0.0.1:8000/",
+  "AI_USERNAME": "admin" ,
```
Comment (cawke) [inline Assets/Resources/MirageXRConfig.json]: you need to put your login password into a separate file. we cannot store private data in a public repository

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+using System.Threading.Tasks;
+using Microsoft.MixedReality.Toolkit.Utilities;
+using Newtonsoft.Json;
+using UnityEngine;
+using UnityEngine.Networking;
+using StringReader = System.IO.StringReader;
+
+namespace MirageXR
+{
+    public class AiServices
```
Comment (cawke) [inline Assets/MirageXR/Scripts/AiServices/AiServices.cs]: you wrote a service. now you need to write a manager that will initialize it (read the config and receive the login password) and manage this service.

File: Assets/MirageXR/Player/Scripts/RootObject.cs [hunk 14]
```diff
@@ -24,6 +24,7 @@ public class RootObject : MonoBehaviour
         private MoodleManager _moodleManager;
         private EditorSceneService _editorSceneService;
         private WorkplaceManager _workplaceManager;
+        private AiServices _aiServices;
```
Comment (cawke) [inline Assets/MirageXR/Player/Scripts/RootObject.cs]: there must be a manager who will manage this service

File: Assets/MirageXR/ContentTypes/Character/Characters/ImageDisplayFiles/MagicParticle.mat [hunk 15]
```diff
@@ -2,14 +2,15 @@
 %TAG !u! tag:unity3d.com,2011:
```
Comment (cawke) [inline Assets/MirageXR/ContentTypes/Character/Characters/ImageDisplayFiles/MagicParticle.mat]: I think this change was made by mistake. roll it back

File: Assets/AddressableAssetsData/AssetGroups/Glyphs.asset [hunk 16]
```diff
@@ -133,6 +133,10 @@ MonoBehaviour:
     m_Address: Glyphs/Unscrew
     m_ReadOnly: 0
     m_SerializedLabels: []
+  - m_GUID: cea501bde7c9ce54dacac44c0cc2e01e
```
Comment (cawke) [inline Assets/AddressableAssetsData/AssetGroups/Glyphs.asset]: I think this change was made by mistake. roll it back

File: Packages/packages-lock.json [hunk 17]
```diff
@@ -31,13 +31,14 @@
     "com.microsoft.mixedreality.openxr": {
```
Comment (cawke) [inline Packages/packages-lock.json]: I think this change was made by mistake. roll it back

File: Assets/XR/Settings/Open XR Package Settings.asset [hunk 18]
```diff
@@ -1,6 +1,6 @@
 %YAML 1.1
 %TAG !u! tag:unity3d.com,2011:
---- !u!114 &-9152892435972783649
```
Comment (cawke) [inline Assets/XR/Settings/Open XR Package Settings.asset]: I think this change was made by mistake. roll it back

**Review by cawke — CHANGES_REQUESTED** [review]

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs (hunk 13 above)
Comment (AKissmehl) [reply Assets/MirageXR/Scripts/AiServices/AiServices.cs]: Okay, can you point me to an existing manager (MirageXR/Managers/?) and explain to me what it should do? I Podbay have to use the unity method like start() awake() and so on. Can we have a chat about what the manager should do?

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs [hunk 19]
```diff
@@ -0,0 +1,479 @@
...
+            {
+                UnityEngine.Debug.LogError($"AudioClip or Model is  null! (Model: {model})");
+            }
+
+
+            List <IMultipartFormSection> fromData = new List<IMultipartFormSection>
+            {
+                new MultipartFormDataSection("model", model),
+                //new MultipartFormFileSection("audio", audioBytArray),
+                new MultipartFormDataSection("dfsdfjk", "dsfsdfkhsdfksdfh"),
```
Comment (cawke) [inline Assets/MirageXR/Scripts/AiServices/AiServices.cs]: I'm guessing it's the remnants of your testing?

**Review by cawke — CHANGES_REQUESTED** [review]

File: Assets/MirageXR/Scripts/AiServices/AiServices.cs (hunk 13 above)
Comment (cawke) [reply Assets/MirageXR/Scripts/AiServices/AiServices.cs]: Now your AiServices class combines the work of both manager and service. in our case service should provide convenient access to work with your server (for example, your methods "Listen", "Think", "Speak" and "AuthenticateUser", although I would change the names). The manager should prepare (initialize) this service: read the keys (AiApiUrl, AiUsername, AiPassword) (we can't store these keys in the public repository, so put them in a file in the "Resources" folder), create an instance of the service, get and store the authentication token. Also all work with the service should be done through the manager, which means that in the manager you need to add methods "Listen", "Think", "Speak" inside which you will check the availability of the token, call the methods of the service and handle possible errors.

example https://github.com/WEKIT-ECS/MIRAGE-XR/blob/release_2.6/Assets/MirageXR/Scripts/OpenAI/OpenAIManager.cs

**Comment by AKissmehl** [comment]
@cawke 
I did not follow the result of our discussion 1 to 1, since I found a simpler solution (at least I think) which needs just an AI Model class and not 3 specific ones. That is the description for the last commit:
The AIManager and AI services have been refactored with a focus on how the Endpoint object will be deserialized.  A class that olds the Token and an AI Model has been added. The AI model will be held in a List for each endpoint and can be reached over a get method.

Old configuration files have been removed, and an error log has been added if the Authentication fails.

Open to discuss:
public async Task<List<OptionsResponse>> GetModelsAsync()
 is, from my point of view, obsolete and can be removed.
 
 I have one Issue, the 11
[Assets/MixedRealityToolkit.Generated/ProjectPreferences.asset](https://github.com/WEKIT-ECS/MIRAGE-XR/pull/1911/files#diff-eaddd58c6f4b03814db0457cb5785ebee3c3f724799ea69d5456b69465712422) is a copy past version out of develop. It still shows up in the PR and I don't know how to remove it.

File: Assets/MirageXR/Managers/AIManager/AIManager.cs [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+        {
+            
+            foreach (var model in endpointModels)
+            {
+                AddModelBasedOnEndpointName(model);
+                UnityEngine.Debug.Log("Model "+model);
+            }
+            UnityEngine.Debug.Log("Set models done! LLM:"+_llmModels.Count+" TTS:"+_ttsModels.Count+ "STT:"+_sttModels.Count );
+
+        }
```
Comment (cawke) [inline Assets/MirageXR/Managers/AIManager/AIManager.cs]:
```suggestion
        private void SetModels(List<AIModel> endpointModels)
        {
            foreach (var model in endpointModels)
            {
                AddModelBasedOnEndpointName(model);
            }
        }
```

File: Assets/MirageXR/Managers/AIManager/AIManager.cs [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+                _models ??= await AiServices.GetAvailableModelsAsync(_url, _token);
+                return _models;
+            }
+            catch (Exception e)
+            {
+                AppLog.LogError(e.ToString());
+                return null;
+            }
+        }
+         */
```
Comment (cawke) [inline Assets/MirageXR/Managers/AIManager/AIManager.cs]:
```suggestion
```

File: Assets/MirageXR/Managers/AIManager/AIManagerDataModel/Token.cs [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+        public Token(string text)
+        {
+            if (string.IsNullOrEmpty(text))
+            {
+                UnityEngine.Debug.LogError("Text cannot be null or empty");
+            }
+            BackendToken = text;
+        }
+    }
+}
```
Comment (cawke) [inline Assets/MirageXR/Managers/AIManager/AIManagerDataModel/Token.cs]:
```suggestion
namespace MirageXR.AIManagerDataModel
{
    public class Token
    {
         [JsonProperty("token")] public string Token { get; set; }
    }
}
```

File: Assets/MirageXR/Managers/AIManager/AiServices.cs [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+            };
+            using var webRequest = UnityWebRequest.Post(apiURL, fromData);
+            webRequest.SetRequestHeader("Authorization", $"Token {token}");
+            await webRequest.SendWebRequest();
+
+            if (webRequest.result != UnityWebRequest.Result.Success)
+            {
+                UnityEngine.Debug.LogError(
+                    $"Error while receiving the result of the Think endpoint: {webRequest.error}");
+            }
```
Comment (cawke) [inline Assets/MirageXR/Managers/AIManager/AiServices.cs]:
```suggestion
            {
                throw new HttpRequestException(
                    $"Error while receiving the result of the Think endpoint: {webRequest.error}");
            }
```

File: Assets/MirageXR/Managers/AIManager/AiServices.cs [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+            {
+                var myDeserializedClass = JsonConvert.DeserializeObject<List<AIModel>>(optionsText);
+                return myDeserializedClass;
+            }
+            catch (Exception e)
+            {
+                UnityEngine.Debug.LogError("Error fetching or deserializing data: " + e.Message);
+                return null;
+            }
+        }
```
Comment (cawke) [inline Assets/MirageXR/Managers/AIManager/AiServices.cs]:
```suggestion
            catch (Exception e)
            {
                throw new HttpRequestException("Error fetching or deserializing data: " + e.Message);
            }
        }
```

File: Assets/MirageXR/Managers/AIManager/AiServices.cs [hunk 25]
```diff
@@ -0,0 +1,232 @@
...
+            var apiURL = $"{url}/speak/";
+            using var webRequest = UnityWebRequestMultimedia.GetAudioClip(apiURL, AudioType.MPEG);
+            webRequest.SetRequestHeader("Authorization", $"Token {token.BackendToken}");
+            webRequest.SetRequestHeader("message", message);
+            webRequest.SetRequestHeader("model", model);
+            await webRequest.SendWebRequest();
+            if (webRequest.result != UnityWebRequest.Result.Success)
+            {
+                UnityEngine.Debug.LogError($"Error while receiving the result of the Speak endpoint: {webRequest.error} {webRequest.result}");
+            }
```
Comment (cawke) [inline Assets/MirageXR/Managers/AIManager/AiServices.cs]:
```suggestion
            if (webRequest.result != UnityWebRequest.Result.Success)
            {
                throw new HttpRequestException($"Error while receiving the result of the Speak endpoint: {webRequest.error} {webRequest.result}");
            }
```

File: Assets/MirageXR/Managers/AIManager/AiServices.cs [hunk 26]
```diff
@@ -0,0 +1,232 @@
...
+            {
+                new MultipartFormDataSection("username", user),
+                new MultipartFormDataSection("password", password),
+            };
+
+            var request = UnityWebRequest.Post($"{apiURL}/authentication/", formData);
+            await request.SendWebRequest();
+            if (request.result != UnityWebRequest.Result.Success)
+            {
+                UnityEngine.Debug.LogError($"Error while receiving the token: {request.error}!");
```
Comment (cawke) [inline Assets/MirageXR/Managers/AIManager/AiServices.cs]:
```suggestion
                throw new HttpRequestException($"Error while receiving the token: {request.error}!");
```

File: Assets/MirageXR/Managers/AIManager/AiServices.cs [hunk 27]
```diff
@@ -0,0 +1,232 @@
...
+
+            var text = request.downloadHandler.text;
+            if (text.Length < 13)
+            {
+                UnityEngine.Debug.LogError($"Error while receiving the token! Got '{text}' and that is too short!");
+            }
+
+            text = text.Substring(10, text.Length - 10 - 2);
+            Token token = new Token(text);
+            return token;
```
Comment (cawke) [inline Assets/MirageXR/Managers/AIManager/AiServices.cs]:
```suggestion
            try
            {
                var token = JsonConvert.DeserializeObject<TokenData>(optionsText);
                return token.Token;
            }
            catch (Exception e)
            {
                throw new HttpRequestException("Error fetching or deserializing data: " + e.Message);
            }
```

File: Assets/MirageXR/Managers/AIManager/AiServices.cs [hunk 28]
```diff
@@ -0,0 +1,232 @@
...
+        }
+
+        /// <summary>
+        /// Authenticates the user and returns a token.
+        /// </summary>
+        /// <param name="apiURL">URL of the server</param>
+        /// <param name="user">Username</param>
+        /// <param name="password">Password</param>
+        /// <returns>The authentication token</returns>
+        public static async Task<Token> AuthenticateUserAsync(string apiURL, string user, string password)
```
Comment (cawke) [inline Assets/MirageXR/Managers/AIManager/AiServices.cs]:
```suggestion
        public static async Task<string> AuthenticateUserAsync(string apiURL, string user, string password)
```

File: Assets/MirageXR/Managers/AIManager/AIManager.cs [hunk 29]
```diff
@@ -0,0 +1,194 @@
...
+
+namespace MirageXR
+{
+    public class AIManager
+    {
+        private string _url;
+        private string _username;
+        private string _password;
+
+        private Token _token;
```
Comment (cawke) [inline Assets/MirageXR/Managers/AIManager/AIManager.cs]:
```suggestion
        private string _token;
```

**Review by cawke — CHANGES_REQUESTED** [review]

**Review by cawke — APPROVED** [review]

**Pull request merged**